File: HW1_C/kmeans.c
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>

int K, N, d, MAX_ITER;

typedef struct Cluster
{
    int size;
    double *prevCentroids;
    double *centroids;
} Cluster;

typedef struct Observation
{
    int cluster;
    double *values;
} Observation;

void printCluster(Cluster *cluster);
void addToMean(Cluster *cluster, Observation *obs);
void removeFromMean(Cluster *cluster, Observation *obs);
void addToClosestcluster(Cluster *clusters, Observation * obs, int first_insert);
double calculateDistance(Observation *observation, Cluster *cluster);
void mainLogic (Cluster* cluster);
int changeMean(Cluster *cluster);

void printCluster(Cluster *cluster)
{
    int i;
    for (i=0; i<d; ++i)
    {
        if (i==d-1)
            printf("%.2f",cluster->prevCentroids[i]);
        else
            printf("%.2f,",cluster->prevCentroids[i]);
    }
}
void addToMean(Cluster *cluster, Observation *obs) {
    int i;
    double sumObs;
    for (i=0; i<d; ++i)
    {
        sumObs = (cluster->centroids[i]) * (cluster ->size);
        cluster->centroids[i] = (sumObs + obs->values[i]) / ((cluster->size)+1);
    }
}

void removeFromMean(Cluster *cluster, Observation *obs)
{
    int i;
    double sumObs;
    for (i=0; i<d; ++i)
    {
        sumObs = (cluster->centroids[i]) * (cluster ->size);
        cluster->centroids[i] = (sumObs - obs->values[i]) / ((cluster->size)-1);
    }
}
int changeMean(Cluster *cluster)
{
    int changed = 0;
    int i;
    for (i=0; i<d; ++i)
    {
        if (cluster->prevCentroids[i] != cluster->centroids[i])
            changed = 1;
        cluster->prevCentroids[i] = cluster->centroids[i];
    }
    return changed;
}

double calculateDistance(Observation *observation, Cluster *cluster)
{
    int i;
    double distance;
    double sum = 0.0f;
        for (i=0; i<d; ++i)
        {
            distance = (cluster->prevCentroids[i] - observation->values[i]);
            distance = distance * distance;
            sum += distance;
        }
    return sum;
}


void addToClosestcluster(Cluster *clusters_arr, Observation *obs, int first_insert) {
    int i, myIndex, prevIndex;
    double min, distance;
    min = calculateDistance(obs, clusters_arr);
    myIndex = 0;
    for (i = 1; i < K; ++i) {
        distance = calculateDistance(obs, &clusters_arr[i]);
        if (distance < min) {
            min = distance;
            myIndex = i;
        }
    }
    prevIndex = (*obs).cluster;
    if (first_insert) {
        (*obs).cluster = myIndex;
        addToMean(clusters_arr +myIndex, obs);
        clusters_arr[myIndex].size += 1;
    } else if (prevIndex != myIndex) {
        (*obs).cluster = myIndex;
        addToMean(clusters_arr+myIndex, obs);
        clusters_arr[myIndex].size += 1;
        removeFromMean(clusters_arr+prevIndex, obs);
        clusters_arr[prevIndex].size -= 1;
    }
}

 void mainLogic (Cluster *clusters_arr)
{
    int counter = 0;
    double n1;
    char c;
    int i = 0;
    int iter = 1;
    int anyoneChanged;
    int j;
    Observation* observations_arr = malloc(N * sizeof(Observation));
    assert(observations_arr != NULL);
    while (scanf("%lf%c", &n1, &c) == 2)
    {
        if (i==0) {
            observations_arr[counter].values = (double*) malloc(d * sizeof(double));
            assert(observations_arr[counter].values != NULL);
        }
        observations_arr[counter].values[i] = n1;
        if (counter<K) {
            if (i == 0) {
                clusters_arr[counter].size = 1;
                observations_arr[counter].cluster = counter;
                clusters_arr[counter].centroids = (double *) malloc(d * sizeof(double));
                assert(clusters_arr[counter].centroids != NULL);
                clusters_arr[counter].prevCentroids = (double *) malloc(d * sizeof(double));
                assert(clusters_arr[counter].prevCentroids != NULL);
                clusters_arr[counter].size = 1;
            }
            clusters_arr[counter].centroids[i] = n1;
            clusters_arr[counter].prevCentroids[i] = n1;
        }
        i += 1;
        if (i == d) {
            i = 0;
            counter += 1;
        }
    }
    for (j=K; j<N; ++j) {
        addToClosestcluster(clusters_arr, &observations_arr[j], 1);
    }
    for(j=0; j<K; ++j)
    {
        changeMean(&clusters_arr[j]);
    }
    while (iter < MAX_ITER) {
        anyoneChanged = 0;
        for (j=0; j<N; ++j) {
            addToClosestcluster(clusters_arr, &observations_arr[j], 0);
        }
        for(j=0; j<K; ++j)
        {
            int changed = changeMean(&clusters_arr[j]);
            if (changed)
                anyoneChanged = 1;
        }
        if (!anyoneChanged)
            break;
        iter += 1;
    }
    for (i=0; i<N; ++i)
    {
        free(observations_arr[i].values);
    }
    free(observations_arr);
}

int main(int argc, char* argv[])
{
    int i;
    Cluster *clusters_arr;
    assert(argc>=5);
    K = atoi(argv[1]);
    N = atoi(argv[2]);
    d = atoi(argv[3]);
    MAX_ITER = atoi(argv[4]);
    if (K==0 || N==0 || d==0 || MAX_ITER==0)
    {
        printf("Not all required arguments were passed\n");
        exit(-1);
    }
    else if (K > N) {
        printf("Cant have more clusters than observations\n");
        exit(-1);
    }

    clusters_arr = malloc(K * sizeof(Cluster));
    mainLogic(clusters_arr);
    for (i=0; i<K; ++i)
    {
        printCluster(&clusters_arr[i]);
        if (i!=K)
            printf("\n");
    }
    for (i=0; i<K; ++i)
    {
        free(clusters_arr[i].centroids);
        free(clusters_arr[i].prevCentroids);
    }
    free(clusters_arr);
    return 0;
}

File: HW1_Python/kmeans.py
import argparse

class Cluster:
    def __init__(self, dValues):
        self.size = 0
        self.centroids = [float(value) for value in dValues]
        self.prevCentroids = [float(value) for value in dValues]

    def __repr__(self):
        tmp = ["%.2f" % value for value in self.centroids]
        return ','.join(tmp)

class Observation:
    def __init__(self, dValues, Cluster):
        self.values = [float(value) for value in dValues]
        self.cluster = Cluster

def main(K, N, d, MAX_ITER):
    observations_arr = []
    clusters = []
    counter = 0
    while True:
        try:
            for line in input().split('\n'):
                values = line.split(',')
                # observation = Observation(values, None)
                # observations_arr.append(observation)
                if counter < K:
                    observation = Observation(values, counter)
                    observations_arr.append(observation)
                    cluster = Cluster(values)
                    clusters.append(cluster)
                    cluster.size += 1
                    # observations_arr[counter].cluster = counter
                    counter += 1
                else:
                    observation = Observation(values, None)
                    observations_arr.append(observation)
        except EOFError:
            break
    for i in range(K, N):
        addToClosestcluster(i, observations_arr, clusters, d, K)
    for indexCluster in range(K):
        changeMean(indexCluster, clusters, d)

    iter = 1
    while iter < MAX_ITER:
        is_changed = False
        for obsNum in range(N):
            addToClosestcluster(obsNum, observations_arr, clusters, d, K)
        for indexCluster in range(K):
            changed = changeMean(indexCluster, clusters, d)
            if changed:
                is_changed = True
        if not is_changed:
            break
        iter += 1
    return (clusters)

def calculateDistance(observation, cluster, d):
    sum = 0
    for indexD in range(d):  # for each centroid in the cluster we will calculate the distance
        distance = (cluster.prevCentroids[indexD] - observation.values[indexD])
        distance *= distance
        sum += distance
    return sum

def addToClosestcluster(obsNum,observations_arr, clusters, d, K):
    # changes = False
    min = calculateDistance(observations_arr[obsNum], clusters[0], d)
    myIndex = 0
    for indexCluster in range(1, K): # for each cluster
        distance = calculateDistance(observations_arr[obsNum], clusters[indexCluster], d)
        if distance < min: # editing the nearest cluster, saving it's index and sum
            min = distance
            myIndex = indexCluster
    prevIndex = observations_arr[obsNum].cluster
    if prevIndex is None:
        observations_arr[obsNum].cluster = myIndex
        addToMean(clusters[myIndex], obsNum, observations_arr, d)
        clusters[myIndex].size += 1
    elif prevIndex != myIndex:
        #if clusters[prevIndex].size() != 1: # if that is the only observation in the cluster we dont want to do anything
        addToMean(clusters[myIndex], obsNum, observations_arr, d)
        clusters[myIndex].size += 1 # edit the size of the set
        removeFromMean(clusters[prevIndex], obsNum, observations_arr, d)
        clusters[prevIndex].size -= 1 # remove it from its last cluster
        observations_arr[obsNum].cluster = myIndex # change the pointer in observations_arr to the new cluster's number

def addToMean(cluster, obsNum, observations_arr, d):
    for indexD in range(d):
        sum = cluster.centroids[indexD] * cluster.size
        cluster.centroids[indexD] = (sum+observations_arr[obsNum].values[indexD]) / (cluster.size+1)

def removeFromMean(cluster, obsNum, observations_arr, d):
    for indexD in range(d):
        sum = cluster.centroids[indexD] * cluster.size
        cluster.centroids[indexD] = (sum - observations_arr[obsNum].values[indexD]) / (cluster.size-1)

def changeMean(indexCluster, clusters, d):
    changed = False
    for indexD in range(d):
        if (clusters[indexCluster].prevCentroids[indexD] != clusters[indexCluster].centroids[indexD]):
            changed = True
        clusters[indexCluster].prevCentroids[indexD] = clusters[indexCluster].centroids[indexD]
    return changed

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("K", help="number of clusters")
    parser.add_argument("N", help="number of observations")
    parser.add_argument("d", help="number of values in an observation")
    parser.add_argument("MAX_ITER", help="max iterations")
    args = parser.parse_args()

    if args.K is None or args.N is None or args.d is None or args.MAX_ITER is None:
        print('Not all required arguments were passed')
        quit()
    K = int(args.K)
    N = int(args.N)
    d = int(args.d)
    MAX_ITER = int(args.MAX_ITER)
    if K <= 0 or N <= 0 or d <= 0 or MAX_ITER <= 0:
        print('Not all required arguments are valid')
        exit(1)
    elif K > N:
        print('Can\'t have more clusters than observations')
        exit(1)
    else:
        results = main(int(args.K), int(args.N), int(args.d), int(args.MAX_ITER))

    for cluster in results:
        print(cluster)